pkg/validationfile/loader.go

```diff
@@ -52,124 +51,96 @@ func PopulateFromFiles(ds datastore.Datastore, filePaths []string) (*PopulatedVa
-			err := rwt.WriteCaveats(ctx, caveatDefs)
+		// Validate and write the object definitions.
+		for _, objectDef := range objectDefs {
+			ts, err := namespace.NewNamespaceTypeSystem(objectDef,
+				namespace.ResolverForDatastoreReader(rwt).WithPredefinedElements(namespace.PredefinedElements{
+					Namespaces: objectDefs,
+				}))
 			if err != nil {
 				return err
 			}
 
-			// Write the object definitions.
-			for _, nsDef := range nsDefs {
-				ts, err := namespace.NewNamespaceTypeSystem(nsDef,
-					namespace.ResolverForDatastoreReader(rwt).WithPredefinedElements(namespace.PredefinedElements{
-						Namespaces: nsDefs,
-					}))
-				if err != nil {
-					return err
-				}
-
-				ctx := dsctx.ContextWithDatastore(ctx, ds)
-				vts, terr := ts.Validate(ctx)
-				if terr != nil {
-					return terr
-				}
-
-				aerr := namespace.AnnotateNamespace(vts)
-				if aerr != nil {
-					return aerr
-				}
-
-				// Write the namespaces.
-				log.Info().Str("filePath", filePath).Str("namespaceName", nsDef.Name).Msg("Loading namespace")
-				if err := rwt.WriteNamespaces(ctx, nsDef); err != nil {
-					return fmt.Errorf("error when loading namespace %s: %w", nsDef.Name, err)
-				}
+			ctx := dsctx.ContextWithDatastore(ctx, ds)
+			vts, terr := ts.Validate(ctx)
+			if terr != nil {
+				return terr
 			}
-			return nil
-		})
-		if lnerr != nil {
-			return nil, revision, lnerr
-		}
 
-		// Load the validation tuples/relationships.
-		var updates []*core.RelationTupleUpdate
-		seenTuples := map[string]bool{}
-		for _, rel := range parsed.Relationships.Relationships {
-			tpl := tuple.MustFromRelationship(rel)
-			updates = append(updates, tuple.Create(tpl))
-			tuples = append(tuples, tpl)
-			seenTuples[tuple.String(tpl)] = true
-		}
-
-		log.Info().Str("filePath", filePath).Int("tupleCount", len(updates)+len(parsed.ValidationTuples)).Msg("Loading test data")
-		for index, validationTuple := range parsed.ValidationTuples {
-			tpl := tuple.Parse(validationTuple)
-			if tpl == nil {
-				return nil, datastore.NoRevision, fmt.Errorf("error parsing validation tuple #%v: %s", index, validationTuple)
+			aerr := namespace.AnnotateNamespace(vts)
+			if aerr != nil {
+				return aerr
 			}
 
-			_, ok := seenTuples[tuple.String(tpl)]
-			if ok {
-				continue
+			if err := rwt.WriteNamespaces(ctx, objectDef); err != nil {
+				return fmt.Errorf("error when loading object definition %s: %w", objectDef.Name, err)
 			}
-			seenTuples[tuple.String(tpl)] = true
-
-			tuples = append(tuples, tpl)
-			updates = append(updates, tuple.Create(tpl))
 		}
 
-		wrevision, terr := ds.ReadWriteTx(ctx, func(rwt datastore.ReadWriteTransaction) error {
-			return rwt.WriteRelationships(ctx, updates)
-		})
-		if terr != nil {
-			return nil, datastore.NoRevision, fmt.Errorf("error when loading validation tuples from file %s: %w", filePath, terr)
+		err = relationships.ValidateRelationshipUpdates(ctx, rwt, updates)
+		if err != nil {
+			return err
```

Comment:
Isn't this something that should happen at the beginning of the transaction? You would have written caveats before validating it. Yes, eventually TX would rollback, but if feels like validation should happen first in order to fail fast?